File: Apps/Flutter/lib/src/rendering/action_conditions.dart
part of '../rendering.dart';

bool isActionVisible(ActionSpec action, RenderContext context) =>
    action.visibleWhen
        .every((condition) => conditionMatches(condition, context));

String? disabledReason(
  ActionSpec action,
  RenderContext context, {
  String fallback = 'This action is not available.',
}) {
  if (!action.disabledWhen
      .any((condition) => conditionMatches(condition, context))) {
    return null;
  }
  return action.disabledTooltip == null
      ? fallback
      : interpolate(action.disabledTooltip!, context);
}

bool conditionMatches(ActionConditionSpec condition, RenderContext context) {
  final value = (contextValue(context, condition.placeholder) ?? '').trim();
  if (condition.exists != null && condition.exists != value.isNotEmpty) {
    return false;
  }
  if (condition.equals != null &&
      value != interpolate(condition.equals!, context)) {
    return false;
  }
  if (condition.notEquals != null &&
      value == interpolate(condition.notEquals!, context)) {
    return false;
  }
  if (condition.inValues.isNotEmpty &&
      !condition.inValues
          .map((item) => interpolate(item, context))
          .contains(value)) {
    return false;
  }
  if (condition.notInValues
      .map((item) => interpolate(item, context))
      .contains(value)) {
    return false;
  }
  if (condition.lessThan != null &&
      !_compareNumeric(value, interpolate(condition.lessThan!, context),
          (left, right) => left < right)) {
    return false;
  }
  if (condition.lessThanOrEqual != null &&
      !_compareNumeric(value, interpolate(condition.lessThanOrEqual!, context),
          (left, right) => left <= right)) {
    return false;
  }
  if (condition.greaterThan != null &&
      !_compareNumeric(value, interpolate(condition.greaterThan!, context),
          (left, right) => left > right)) {
    return false;
  }
  if (condition.greaterThanOrEqual != null &&
      !_compareNumeric(
          value,
          interpolate(condition.greaterThanOrEqual!, context),
          (left, right) => left >= right)) {
    return false;
  }
  return true;
}

ActionPrecheckResult? evaluateActionPrecheck(
    ActionPrecheckSpec? precheck, RenderContext context) {
  final raw = precheck?.diskSpaceGB?.trim();
  if (raw == null || raw.isEmpty) {
    return null;
  }
  final requiredGB = evaluateNumeric(interpolate(raw, context));
  if (requiredGB == null || requiredGB <= 0) {
    return null;
  }
  var targetPath =
      interpolate(precheck!.diskSpacePath ?? '{{out_dir}}', context).trim();
  if (targetPath.isEmpty) {
    targetPath = context.bundleRootPath;
  }
  final resolvedPath = resolveUserPath(targetPath, context.bundleRootPath);
  final availableGB = _volumeAvailableGB(resolvedPath);
  if (availableGB == null) {
    return null;
  }
  final isLow = availableGB < requiredGB;
  final required = _formatGB(requiredGB);
  final available = _formatGB(availableGB);
  final pathLabel = File(resolvedPath).parent.path;
  final message = isLow && precheck.warningMessage != null
      ? interpolate(precheck.warningMessage!, context)
      : isLow
          ? 'Need $required GB free at $pathLabel, only $available GB available.'
          : 'Estimated $required GB needed at $pathLabel ($available GB free).';
  return ActionPrecheckResult(
    severity:
        isLow ? ActionPrecheckSeverity.warning : ActionPrecheckSeverity.info,
    title: isLow ? 'Not enough free disk space' : 'Disk space estimate',
    message: message,
  );
}

bool _compareNumeric(
  String left,
  String right,
  bool Function(double left, double right) compare,
) {
  final leftValue = evaluateNumeric(left);
  final rightValue = evaluateNumeric(right);
  return leftValue != null &&
      rightValue != null &&
      compare(leftValue, rightValue);
}

double? _volumeAvailableGB(String path) {
  if (Platform.isWindows) {
    return null;
  }
  var probe = path;
  while (probe.isNotEmpty &&
      FileSystemEntity.typeSync(probe) == FileSystemEntityType.notFound) {
    final parent = File(probe).parent.path;
    if (parent == probe) {
      break;
    }
    probe = parent;
  }
  final ProcessResult result;
  try {
    result = Process.runSync('df', ['-k', probe]);
  } on ProcessException {
    return null;
  }
  if (result.exitCode != 0) {
    return null;
  }
  final lines = result.stdout.toString().trim().split(RegExp(r'\r?\n'));
  if (lines.length < 2) {
    return null;
  }
  final columns = lines.last.trim().split(RegExp(r'\s+'));
  if (columns.length < 4) {
    return null;
  }
  final availableK = int.tryParse(columns[3]);
  return availableK == null ? null : availableK / 1048576.0;
}

String _formatGB(double value) {
  if (value >= 100) {
    return value.toStringAsFixed(0);
  }
  if (value >= 10) {
    return value.toStringAsFixed(1);
  }
  return value.toStringAsFixed(2);
}

enum ActionPrecheckSeverity { info, warning }

class ActionPrecheckResult {
  const ActionPrecheckResult({
    required this.severity,
    required this.title,
    required this.message,
  });

  final ActionPrecheckSeverity severity;
  final String title;
  final String message;
}

class _NumericParser {
  _NumericParser(this.text);

  final String text;
  int index = 0;

  double? parse() {
    final value = _expression();
    _skipWhitespace();
    return index == text.length ? value : null;
  }

  double? _expression() {
    var value = _term();
    while (value != null) {
      _skipWhitespace();
      if (_consume('+')) {
        final right = _term();
        if (right == null) return null;
        value += right;
      } else if (_consume('-')) {
        final right = _term();
        if (right == null) return null;
        value -= right;
      } else {
        return value;
      }
    }
    return null;
  }

  double? _term() {
    var value = _factor();
    while (value != null) {
      _skipWhitespace();
      if (_consume('*')) {
        final right = _factor();
        if (right == null) return null;
        value *= right;
      } else if (_consume('/')) {
        final right = _factor();
        if (right == null) return null;
        value /= right;
      } else {
        return value;
      }
    }
    return null;
  }

  double? _factor() {
    _skipWhitespace();
    if (_consume('+')) return _factor();
    if (_consume('-')) {
      final value = _factor();
      return value == null ? null : -value;
    }
    if (_consume('(')) {
      final value = _expression();
      return _consume(')') ? value : null;
    }
    return _number();
  }

  double? _number() {
    _skipWhitespace();
    final start = index;
    while (index < text.length && RegExp(r'[0-9.]').hasMatch(text[index])) {
      index += 1;
    }
    return start == index
        ? null
        : double.tryParse(text.substring(start, index));
  }

  bool _consume(String token) {
    if (index < text.length && text[index] == token) {
      index += 1;
      return true;
    }
    return false;
  }

  void _skipWhitespace() {
    while (index < text.length && text[index].trim().isEmpty) {
      index += 1;
    }
  }
}

File: exp-platform/kotlin/compose/shared/src/main/kotlin/dev/guiforcli/compose/runtime/ActionConditions.kt
package dev.guiforcli.compose.runtime

import dev.guiforcli.compose.model.ActionConditionSpec
import dev.guiforcli.compose.model.ActionPrecheckSpec
import dev.guiforcli.compose.model.ActionSpec
import java.io.File

data class ActionPrecheckResult(
    val severity: Severity,
    val title: String,
    val message: String,
) {
    enum class Severity { Info, Warning }
}

fun isActionVisible(action: ActionSpec, context: RenderContext): Boolean =
    action.visibleWhen.all { conditionMatches(it, context) }

fun disabledReason(
    action: ActionSpec,
    context: RenderContext,
    fallback: String = "This action is not available.",
): String? {
    if (action.disabledWhen.none { conditionMatches(it, context) }) {
        return null
    }
    return action.disabledTooltip?.let { interpolate(it, context) } ?: fallback
}

fun conditionMatches(condition: ActionConditionSpec, context: RenderContext): Boolean {
    val value = contextValue(context, condition.placeholder)?.trim().orEmpty()
    if (condition.exists != null && condition.exists != value.isNotEmpty()) {
        return false
    }
    if (condition.equals != null && value != interpolate(condition.equals, context)) {
        return false
    }
    if (condition.notEquals != null && value == interpolate(condition.notEquals, context)) {
        return false
    }
    if (condition.inValues.isNotEmpty() && value !in condition.inValues.map { interpolate(it, context) }) {
        return false
    }
    if (value in condition.notInValues.map { interpolate(it, context) }) {
        return false
    }
    if (condition.lessThan != null && !compareNumeric(value, interpolate(condition.lessThan, context)) { left, right -> left < right }) {
        return false
    }
    if (condition.lessThanOrEqual != null && !compareNumeric(value, interpolate(condition.lessThanOrEqual, context)) { left, right -> left <= right }) {
        return false
    }
    if (condition.greaterThan != null && !compareNumeric(value, interpolate(condition.greaterThan, context)) { left, right -> left > right }) {
        return false
    }
    if (condition.greaterThanOrEqual != null && !compareNumeric(value, interpolate(condition.greaterThanOrEqual, context)) { left, right -> left >= right }) {
        return false
    }
    return true
}

fun evaluateActionPrecheck(precheck: ActionPrecheckSpec?, context: RenderContext): ActionPrecheckResult? {
    val rawRequired = precheck?.diskSpaceGB?.trim().orEmpty()
    if (rawRequired.isEmpty()) {
        return null
    }
    val requiredGB = evaluateNumeric(interpolate(rawRequired, context)) ?: return null
    if (requiredGB <= 0.0) {
        return null
    }
    val targetPath = interpolate(precheck?.diskSpacePath ?: "{{out_dir}}", context)
        .ifBlank { context.bundleRootPath }
    val resolvedPath = resolveUserPath(targetPath, context.bundleRootPath)
    val availableGB = availableGB(resolvedPath) ?: return null
    val isLow = availableGB < requiredGB
    val pathLabel = nearestExistingPath(resolvedPath)
    val message = if (isLow && precheck?.warningMessage != null) {
        interpolate(precheck.warningMessage, context)
    } else if (isLow) {
        "Need ${formatGB(requiredGB)} GB free at $pathLabel; ${formatGB(availableGB)} GB is available."
    } else {
        "Estimated ${formatGB(requiredGB)} GB needed at $pathLabel (${formatGB(availableGB)} GB free)."
    }
    return ActionPrecheckResult(
        severity = if (isLow) ActionPrecheckResult.Severity.Warning else ActionPrecheckResult.Severity.Info,
        title = if (isLow) "Not enough free disk space" else "Disk space estimate",
        message = message,
    )
}

fun evaluateNumeric(expression: String): Double? = NumericParser(expression).parse()

private fun compareNumeric(left: String, right: String, compare: (Double, Double) -> Boolean): Boolean {
    val leftValue = evaluateNumeric(left) ?: return false
    val rightValue = evaluateNumeric(right) ?: return false
    return compare(leftValue, rightValue)
}

private fun availableGB(path: String): Double? {
    val probe = File(nearestExistingPath(path))
    return if (probe.exists()) {
        probe.usableSpace / 1_073_741_824.0
    } else {
        null
    }
}

private fun nearestExistingPath(path: String): String {
    var probe = File(path)
    while (!probe.exists() && probe.parentFile != null) {
        probe = probe.parentFile
    }
    return probe.path
}

private fun formatGB(value: Double): String = when {
    value >= 100 -> "%.0f".format(value)
    value >= 10 -> "%.1f".format(value)
    else -> "%.2f".format(value)
}

private class NumericParser(private val text: String) {
    private var index = 0

    fun parse(): Double? {
        val value = expression()
        skipWhitespace()
        return if (index == text.length) value else null
    }

    private fun expression(): Double? {
        var value = term() ?: return null
        while (true) {
            skipWhitespace()
            value = when {
                consume('+') -> value + (term() ?: return null)
                consume('-') -> value - (term() ?: return null)
                else -> return value
            }
        }
    }

    private fun term(): Double? {
        var value = factor() ?: return null
        while (true) {
            skipWhitespace()
            value = when {
                consume('*') -> value * (factor() ?: return null)
                consume('/') -> value / (factor() ?: return null)
                else -> return value
            }
        }
    }

    private fun factor(): Double? {
        skipWhitespace()
        if (consume('+')) return factor()
        if (consume('-')) return factor()?.unaryMinus()
        if (consume('(')) {
            val value = expression()
            return if (consume(')')) value else null
        }
        return number()
    }

    private fun number(): Double? {
        skipWhitespace()
        val start = index
        while (index < text.length && (text[index].isDigit() || text[index] == '.')) {
            index += 1
        }
        return text.substring(start, index).takeIf { it.isNotEmpty() }?.toDoubleOrNull()
    }

    private fun skipWhitespace() {
        while (index < text.length && text[index].isWhitespace()) {
            index += 1
        }
    }

    private fun consume(expected: Char): Boolean {
        if (index >= text.length || text[index] != expected) {
            return false
        }
        index += 1
        return true
    }
}
